### PR TITLE
CA1853: Address issues found in PR review #6767

### LINKED ISSUE #6781: CA1853: Wrong code fix is suggested when arguments of `ContainsKey` and `Remove` differ
### Analyzer

**Diagnostic ID**: [CA1853](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1853): `Unnecessary call to 'Dictionary.ContainsKey(key)'`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers/8.0.0-preview.23320.2)

**Version**: 8.0.0-preview.23320.2 (Latest)

### Describe the bug

A code fix is suggested if `ContainsKey` is called with a different key than `Remove`. This results in incorrect code.

### Steps To Reproduce

```C#
var dict = new Dictionary<string, string>();
var otherKey = "other";

if (dict.ContainsKey("key"))
    dict.Remove(otherKey);
```

which gets fixed to:

```C#
var dict = new Dictionary<string, string>();
var otherKey = "other";

dict.Remove(otherKey);
```

or as a test case:

```C#
[Fact]
public async Task ContainsAndAddCalledWithDifferentArguments_NoDiagnostic_CS()
{
    string source = CSUsings + CSNamespaceAndClassStart + @"
private readonly Dictionary<string, string> MyDictionary = new Dictionary<string, string>();
private const string KeyField = ""Field Key"";

public string KeyProperty { get; } = ""Property Key"";

public MyClass(string keyParameter)
{
    if (!MyDictionary.ContainsKey(""Key""))
        MyDictionary.Remove(""Other Key"");

    if (!MyDictionary.ContainsKey(""Key""))
        MyDictionary.Remove(keyParameter);

    if (!MyDictionary.ContainsKey(""Key""))
        MyDictionary.Remove(KeyField);

    if (!MyDictionary.ContainsKey(""Key""))
        MyDictionary.Remove(KeyProperty);
    
    string otherItemLocal = ""Other Key"";
    if (!MyDictionary.ContainsKey(""Key""))
        MyDictionary.Remove(otherItemLocal);
}" + CSNamespaceAndClassEnd;

    await VerifyCS.VerifyCodeFixAsync(source, source);
}
```

The above leads to 5 diagnostics.

### Expected behavior

No code fix and no diagnostic is suggested.

### Actual behavior

Code fix is suggested that leads to a different behavior.

## Additional context

Found by @buyaa-n during #6767.

### PR DESCRIPTION
**Pull request by mpidash**
The implementation of CA1868 (`Unnecessary call to 'Contains' for sets`) was based on the analyzer CA1853 (`Unnecessary call to 'Dictionary.ContainsKey(key)`), since they are very similar. During the review of #6767, a couple of problems were found that should also be addressed for CA1853.

This PR merges the analyzer for guarded calls, leading to the following changes for CA1853:

- Fixes #6781
- Support for ternary operators
- Support for guarded calls in the else branch of a conditional
- Do not report a diagnostic when the condition is negated (same behavior as the set analyzer)
- Support for variable assignments
- Support for derived types (e.g. `SortedDictionary<TKey, TValue>` or `ImmutableSortedDictionary<TKey, TValue>.Builder`)
- Support for access through interface types (`IDictionary<TKey, TValue>`)
- Support for `IImmutableDictionary<TKey, TValue>` and its derived types
- Support for guarded calls that are implemented through extension methods (e.g. `System.Collections.Generic.CollectionExtensions.Remove(IDictionary<TKey, TValue>, TKey, out TValue)` for `SortedDictionary<TKey, TValue>` (as `Remove(TKey, out TValue` is part of the concrete type `Dictionary<TKey, TValue>`))

This should also make it easier to add other guarded call diagnostics in the future.

### CODE REVIEW

**Comment by mpidash** [comment]
@Youssef1313 @buyaa-n @sharwell What do you think about combining the analyzers for guarded calls? This would help to apply the feedback for the set analyzer to the dictionary analyzer.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardCall.cs [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            foreach (var guardedCallContext in s_guardedCallContexts)
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardCall.cs:92]: Wonder if you tried/considered to populate all symbols once and register one action to check all, might be more efficient.

**Review by buyaa-n — APPROVED** [review]
@mpidash thank you for fixing this, overall LGTM

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKeyTests.cs [hunk 2]
```diff
@@ -531,72 +2023,376 @@ End Class
...
+                    }
+
+                    private void RemoveItem(Dictionary<string, string> dictionaryParam)
+                    {
+                        if (setParam.ContainsKey("Item"))
+                            DictionaryField1.Remove("Item");
+
+                        if (!DictionaryField1.ContainsKey("Item"))
+                        {
+                            setParam.Remove("Item");
```
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKeyTests.cs:2071]:
```suggestion
                        if (dictionaryParam.ContainsKey("Item"))
                            DictionaryField1.Remove("Item");

                        if (!DictionaryField1.ContainsKey("Item"))
                        {
                            dictionaryParam.Remove("Item");
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardCall.cs (hunk 1 above)
Comment (mpidash) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardCall.cs:92]: I was thinking about this in testing as I had a bug where some guarded calls were found twice (two separate actions found them) but did not follow through.
Should I prepare a new PR where everything is consolidated into one action or should we keep it like this?

Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardCall.cs:92]: > Should I prepare a new PR where everything is consolidated into one action 

That would be great, thank you!

Comment (mpidash) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardCall.cs:92]: I've tested loading all symbols once and then registering one action like this:

```cs
private void OnCompilationStart(CompilationStartAnalysisContext context)
{
    var guardedCallContextWithSymbols = s_guardedCallContexts
        .Select(c => GuardedCallSymbols.TryGetSymbols(context.Compilation, c, out var symbols)
            ? (Context: c, Symbols: symbols)
            : (Context: c, Symbols: null))
        .WhereAsArray(p => p.Symbols is not null);

    if (!guardedCallContextWithSymbols.IsEmpty)
    {
        context.RegisterOperationAction(OnConditional, OperationKind.Conditional);

        void OnConditional(OperationAnalysisContext context)
        {
            foreach (var (guardedCallContext, symbols) in guardedCallContextWithSymbols)
            {
                var conditional = (IConditionalOperation)context.Operation;

                if (!symbols!.HasApplicableConditionInvocation(conditional.Condition, out var conditionInvocation, out bool containsNegated) ||
                    !symbols.HasApplicableGuardedInvocation(conditional, containsNegated, out var guardedInvocation) ||
                    !AreInvocationsOnSameInstance(conditionInvocation, guardedInvocation) ||
                    !AreInvocationArgumentsEqual(conditionInvocation, guardedInvocation))
                {
                    continue;
                }

                using var locations = ArrayBuilder<Location>.GetInstance(2);
                locations.Add(conditional.Syntax.GetLocation());
                locations.Add(guardedInvocation.Syntax.Parent!.GetLocation());

                context.ReportDiagnostic(conditionInvocation.CreateDiagnostic(
                    guardedCallContext.Rule,
                    additionalLocations: locations.ToImmutable(),
                    properties: null,
                    guardedInvocation.TargetMethod.ToDisplayString(s_symbolDisplayFormat),
                    conditionInvocation.TargetMethod.ToDisplayString(s_symbolDisplayFormat)));
            }
        }
    }
}
```

I've tried to measure the performance difference with `AnalyzerRunner`. The variance between runs of the same code (old one and the updated above) was quite high on my PC (from 5s to 11s for both, tested against `dotnet/roslyn` with a Release build), which makes it kinda hard to say for sure that one is more performant than the other.

Do you know a better way to compare the performance of two analyzers?

Comment (buyaa-n) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardCall.cs:92]: > Do you know a better way to compare the performance of two analyzers?

I haven't really tried to measure a perf of analyzers, there is some suggestions in https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Performance.md, you might have already checked that. 

We also have some [PerformanceTests](https://github.com/dotnet/roslyn-analyzers/tree/main/src/PerformanceTests) in the repo, by my understanding that runs the [tests](https://github.com/dotnet/roslyn-analyzers/blob/main/src/PerformanceTests/Tests) on CI, saves the results somewhere and compares them with the next PR's perf build and fails if the diff was above the threshold, not sure if adding a test for the analyzer there and use the old version as a `Base` could be a better way. The [Perf.cmd](https://github.com/dotnet/roslyn-analyzers/blob/main/Perf.cmd) script in the root would run all performance tests (which probably would take several hours to finish), unfortunately it seems there is no way to run only one specific perf test  

Ping @mavasani for better suggestions

**Pull request merged**